group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1759,20 +1766,22 @@ private List<Record> shareGroupFenceMember(
         if (!subscriptionMetadata.equals(group.subscriptionMetadata())) {
             log.info("[GroupId {}] Computed new subscription metadata: {}.",
                 group.groupId(), subscriptionMetadata);
-            records.add(newGroupSubscriptionMetadataRecord(group.groupId(), subscriptionMetadata));
+            replayRecords.add(newGroupSubscriptionMetadataRecord(group.groupId(), subscriptionMetadata));
         }
 
+        // We bump the group epoch.
+        int groupEpoch = group.groupEpoch() + 1;
+        records.add(newGroupEpochRecord(group.groupId(), groupEpoch, SHARE));
+
         cancelTimers(group.groupId(), member.memberId());
 
         /*
          Replay generated records for share consumer as only group epoch records are persisted for
          share consumer.
         */
-        records.forEach(record -> replay(record, SHARE));
+        replayRecords.forEach(record -> replay(record, SHARE));
```

Comment:
What would happen if the real records could not be written? We would have updated the soft state, and not have updated the hard state.

Reply:
ReplayRecords only hold those changes which are not to be persisted. The group epoch is only bumped in the callback post persistence, members are do created and soft state in-memory is updated (as like consumer groups) for which if persistence fails and co-ordinator changes then we will see UNKNOWN_MEMBER error. The behaviour is as like current consumer groups.